alpenhorn/storage.py

```diff
@@ -52,26 +108,129 @@ class StorageNode(base_model):
+    def under_min(self) -> bool:
+        """Is the amount of free space below the minimum allowed?
+
+        Returns False if `avail_gb` is None.
+        """
+        if self.avail_gb is None:
+            return False
+        return self.avail_gb < self.min_avail_gb
+
+    def over_max(self) -> bool:
+        """Is the total size of file copies on the node greater than allowed?
+
+        If max_total_gb is non-positive or None, returns False.
+        """
+        if self.max_total_gb is None or self.max_total_gb <= 0:
+            return False
+        return self.total_gb() >= self.max_total_gb
+
+    def named_copy_present(self, acqname: str, filename: str) -> bool:
+        """Is an ArchiveFileCopy named `acqname/filename` present?
+
+        Both the file and the acquisition must exist in the database.
+
+        Parameters
+        ----------
+        acqname : str
+            The name of the ArchiveAcq
+        filename : str
+            The name of the ArchiveFile
+        """
+        from .archive import ArchiveFileCopy
+        from .acquisition import ArchiveFile, ArchiveAcq
+
+        try:
+            ArchiveFileCopy.select().join(ArchiveFile).join(ArchiveAcq).where(
+                ArchiveAcq.name == acqname,
+                ArchiveFile.name == filename,
+                ArchiveFileCopy.node == self,
+                ArchiveFileCopy.has_file == "Y",
+            ).get()
+        except pw.DoesNotExist:
+            return False
+
+        return True
+
+    def copy_present(self, file: ArchiveFile) -> bool:
+        """Is a copy of ArchiveFile `file` present on this node?"""
+        from .archive import ArchiveFileCopy
+
+        try:
+            ArchiveFileCopy.get(
+                ArchiveFileCopy.file == file,
+                ArchiveFileCopy.node == self,
+                ArchiveFileCopy.has_file == "Y",
+            )
+        except pw.DoesNotExist:
+            return False
+
+        return True
+
+    def total_gb(self) -> float:
+        """The size in GiB of all files on node
+
+        This is the total of all apparent file sizes from the database.
+        It may be quite different than the amount of actual space the
+        file copies take up on the underlying storage system.
+        """
+        from .acquisition import ArchiveFile
+        from .archive import ArchiveFileCopy
+
+        size = (
+            ArchiveFile.select(fn.Sum(ArchiveFile.size_b))
+            .join(ArchiveFileCopy)
+            .where(ArchiveFileCopy.node == self, ArchiveFileCopy.has_file == "Y")
+        ).scalar(as_tuple=True)[0]
+
+        return 0.0 if size is None else float(size) / 2**20
+
+    def all_files(self) -> list[pathlib.PurePath]:
+        """A list of all files on node
+
+        Returns
+        -------
+        all_files : list
+            a list of paths relative to `root` for all file copies
+            exsiting on the node.
+        """
+        from .archive import ArchiveFileCopy
+
+        return [
```

Comment:
Should return a set, not a list, because we're going to be doing `path in all_files()` on this output, not iterating through it.